# ENGR 418 Project Stage 1


## Group 5


Jason Schultz - 24662652


Matthew Hinchliff - 97352314

In [24]:
'''
Sets up all library dependencies as well all required function calls
to easily fit the classifiers to a lego shape data set.
'''

# Import necessary libraries for data processing and model training
import os
import numpy as np
from numpy import ndarray
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (accuracy_score, confusion_matrix)
from PIL import Image

        
def get_matching_image_files(folder: str, label: str) -> list[str]:
    '''
    Reads in all images at a given folder path and returns all files
    that contain the label string in it.
    '''
    # Get the list of all files in the folder
    file_names = os.listdir(folder)
    
    # Filter files based on the label prefix
    label_files = []
    for file in file_names:
        if file.startswith(label):
            label_files.append(file)
    
    return label_files

        
def read_image_as_array(im_width: int, filepath: str) -> ndarray[float]:
    '''
    Takes in a path to an image and resizes and returns it to a square 
    image based on given im_width value.
    '''
    im = Image.open(filepath).convert("L")  # Convert to grayscale
    im = im.resize((im_width, im_width))  # Resize to target width
    im_array = np.asarray(im)  # Convert image to array

    return im_array


def get_classifier_data(
        folder: str,
        im_width: int,
        label: str,
        label_index: int,
) -> tuple[ndarray, ndarray]:
    ''' 
    Reads in image data for a single classifier and returns its feature array
    and label array
    '''
    label_files = get_matching_image_files(folder, label)

    # Initialize empty arrays for storing image data and labels
    x = np.empty((len(label_files), im_width**2))  # Features
    y = np.empty((len(label_files), 1))            # Labels

    # Read images, resize, and store them in x; assign corresponding label to y
    for i in range(len(label_files)):
        filepath = folder + label_files[i]
        im_array = read_image_as_array(im_width, filepath)
        x[i:] = im_array.reshape(1, -1)  # Flatten image array
        y[i, 0] = label_index  # Assign label index to the output array

    return x, y  # Return feature and label arrays


def get_feature_and_label_arrays(
        path: str,
        labels: list[str],
        im_width: int,
) -> tuple[ndarray, ndarray]:
    ''' 
    Obtains the feature matrix and label vector for all classifiers and concatenates
    them all to a single array, x and y.
    '''
    # Initialize lists to store the training data
    x_list = []
    y_list = []

    # Loop through each label to load data, using the get_data function defined earlier
    for (i, label) in enumerate(labels):
        x_i, y_i = get_classifier_data(path, im_width, label, i)  # Get data for each label
        x_list.append(x_i)  # Add features (x) to the list
        y_list.append(y_i)  # Add labels (y) to the list

    # Initialize x_train and y_train with the first label's data
    x = np.array(x_list[0])
    y = np.array(y_list[0])

    # Append data from the remaining labels to x_train and y_train
    for i in range(1, len(x_list)):
        x = np.append(x, x_list[i], axis=0)
        y = np.append(y, y_list[i], axis=0)

    return x, y


def test_function(
        path: str,
        model: LogisticRegression,
        labels: list[str],
        im_width: int,
) -> None:
    ''' 
    Predicts the classifiers of a given data set in path for the model and
    prints out the confusion matrix and accuracy score
    '''
    x, y = get_feature_and_label_arrays(path, labels, im_width)
    y_pred = model.predict(x)
    evaluate_model_metrics(y, y_pred)


def evaluate_model_metrics(true_labels, predicted_labels):
    ''' 
    Function to print the confusion matrix and accuracy score to evaluate model performance
    '''
    print("Confusion Matrix:")
    print(confusion_matrix(true_labels, predicted_labels))
    print("\nAccuracy Score:")
    print(accuracy_score(true_labels, predicted_labels))


In [22]:
'''
Training cell

This cell sets up the required variables to run the test as well as fits
the model to the training data set. A test is also performed to see its
accuracy on the trained data.
'''
# Initializing required parameters
path = "../project/Lego_dataset_1/training/"
im_width = 64
labels = ["2b1", "cir", "rec", "squ"]  # List of strings to separate images based on classifier
model = LogisticRegression(max_iter = 1000)

# Fitting the model on training data
x_train, y_train = get_feature_and_label_arrays(path, labels, im_width)
model.fit(x_train, y_train)

# Running test on training data to see its accuracy score
test_function(path, model, labels, im_width)

/home/jason/coding_projects/python/engr418/.venv/lib/python3.10/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Confusion Matrix:
[[36  0  0  0]
 [ 0 36  0  0]
 [ 0  0 36  0]
 [ 0  0  0 36]]
Accuracy Score:
1.0


In [25]:
'''
Testing cell.

This cell changes the path to the testing data set and predicts the results 
on the testing set to compare its accuracy score with the training data.
'''
path = "../project/Lego_dataset_1/testing/"
test_function(path, model, labels, im_width)

Confusion Matrix:
[[18  0  0  0]
 [ 0 17  0  1]
 [ 0  0 17  1]
 [ 0  1  0 17]]

Accuracy Score:
0.9583333333333334


# Results

| im_width | Weights | Training Accuracy | Testing Accuracy |
| -------- | ------- | ----------------- | ---------------- |
| 8        |  65     | 1.0               | 0.8611           |
| 16       |  257    | 1.0               | 0.9306           |
| 32       |  1025   | 1.0               | 0.9444           |
| 64       |  4097   | 1.0               | 0.9583           |